[SV-COMP'18 12/19] SV-COMP graphml fixes [blocks: #3486]

tautschnig opened the pull request:
Do not merge as-is: this needs cleanup of patches and of the changes to `expr2ct`.

peterschrammel committed:
No outgoing edges from violation nodes

The witness format documentation only states "Nodes where this flag is
set must not have any leaving transitions" for sink nodes, but quite
possibly validators expect the same for violation nodes.

marek-trtik committed:
Using cache to speed up generation of GraphML file

Expression-to-string conversion is costly.

peterschrammel committed:
GraphML: add cyclehead node

It's actually not clear whether this is necessary or permitted, it's not
contained in the official specification.

peterschrammel committed:
Remove obsolete graphml frontierNode, nodeType

See https://github.com/sosy-lab/sv-witnesses for the most up-to-date
format specification.

peterschrammel committed:
Remove obsolete sourcecode element in graphml

https://github.com/sosy-lab/sv-witnesses documents a "control" node, and
examples still include "sourcecode" nodes, but there neither is
documented semantics for "control" nodes nor is it clear whether the
"sourcecode" node is used by any tool.

peterschrammel committed:
GraphML output: hide assignments to dynamic objects

These symbols do not occur in the input program.

peterschrammel committed:
GraphML output: Split array-list into assignments

Witness validators cannot handle array lists.

marek-trtik committed:
GraphML: Use the \result symbol in place of return value symbols

The witness specification permits use of \result in assumptions (in
violation witnesses), but then also requires setting
assumption.resultfunction.

lucasccordeiro committed:
Add violation node for the memory safety benchmarks

We must not filter out generated assertions for memory leaks.

tautschnig committed:
expr2ct/clean configuration: Do not output C library macros

NULL, NAN, INFINITY are not a keywords, just macros. Do not use them for
output under the "clean" configuration.

smowton commented on src/ansi-c/expr2c.cpp:
```diff
@@ -699,6 +699,7 @@ std::string expr2ct::convert_struct_type(
   if(tag!="")
     dest+=" "+id2string(tag);
 
+#if 0
```
Add comment explaining why?

smowton commented on src/goto-programs/graphml_witness.cpp:
```diff
@@ -52,9 +57,38 @@ std::string graphml_witnesst::convert_assign_rec(
   std::string result;
 
-  if(assign.rhs().id()==ID_array)
+  if(assign.rhs().id()=="array-list")
```
Use ID, or add one if not already present

smowton commented on src/goto-programs/graphml_witness.cpp:
```diff
@@ -129,13 +163,17 @@ std::string graphml_witnesst::convert_assign_rec(
+    exprt clean_lhs=assign.lhs();
+    remove_l0_l1(clean_lhs);
+    std::string lhs=from_expr(ns, identifier, clean_lhs);
+    if(lhs.find("#return_value")!=std::string::npos ||
```
Use ID

smowton commented on src/goto-programs/graphml_witness.cpp:
```diff
@@ -169,22 +207,86 @@ static bool filter_out(
+  return false;
+}
+
+static bool contains_symbol_prefix(const exprt &expr, const std::string &prefix){
```
Use `depth_begin` et al instead of recursing?

smowton commented on src/goto-programs/graphml_witness.cpp:
```diff
@@ -169,22 +207,86 @@ static bool filter_out(
+    {
+      nodes.push_back(graphml.add_node());
+      graphml[nodes.back()].node_name="N" + std::to_string(i);
+      graphml[nodes.back()].is_violation=(i==max_thread_idx+1U) ? true : false;
```
I think that ternary conditional is redundant

smowton commented on src/goto-programs/graphml_witness.cpp:
```diff
@@ -169,22 +207,86 @@ static bool filter_out(
+      graphml[nodes.back()].has_invariant=false;
+    }
+
+    for(auto it=nodes.cbegin(); std::next(it)!=nodes.cend(); ++it)
```
`INVARIANT(!nodes.empty())`?

smowton commented on src/goto-programs/graphml_witness.cpp:
```diff
@@ -275,16 +381,38 @@ void graphml_witnesst::operator()(const goto_tracet &goto_trace)
+          }
+          else if(
+            id2string(it->lhs_object.get_identifier())
+              .find("#return_value")==std::string::npos &&
```
Several well-known identifiers to ID-ify here

smowton commented on src/goto-programs/graphml_witness.cpp:
```diff
@@ -300,33 +428,13 @@ void graphml_witnesst::operator()(const goto_tracet &goto_trace)
                 it->pc->is_goto())
         {
+#if 0
+          // TODO: this requires knowledge about the phase of the
```
Provide an issue reference?

smowton commented on src/util/std_pair_hash.h:
```diff
@@ -0,0 +1,35 @@
+inline void hash_combine(std::size_t & seed, const T & v)
+{
+  std::hash<T> hasher;
+  seed ^= hasher(v) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
```
Provide a reference for the particular source of magic numbers?

tautschnig committed:
Use clean-C configuration for GraphML witnesses

Witness validators expect syntactically correct C code, not
human-friendly pseudo-C.

tautschnig committed:
GraphML witnesses: unwind "with" expressions

Witness validators cannot be expected to understand CBMC's "with"
expressions.

Fixes: #4486

tautschnig committed:
Make sure SSA suffixes are stripped off GraphML output

Witness validators expect syntactically correct C code, and cannot cope
with SSA suffixes.

tautschnig committed:
Hack to make CPAchecker parse graphml concurrency witnesses

This does likely need more work, but at least got us some points in the
past.

tautschnig committed:
GraphML witnesses: clang-format switch/case statement

tautschnig committed:
GraphML witnesses: correctly set the scope of parameters

We would previously use the name of the calling function. Instead,
parameter assignments require a special case. Regression test kindly
provided by @JanSvejda.

Fixes: #4418

tautschnig commented:
This PR is now actually ready for review. We certainly have a lot of problems left to resolve, but that would at least get develop to have the GraphML witness code that we've been using in SV-COMP the last years.

smowton commented on src/ansi-c/expr2c.cpp:
```diff
@@ -1866,6 +1869,27 @@ std::string expr2ct::convert_constant(
       else if(dest=="-NaN")
         dest="-NAN";
     }
+    else if(dest.size() == 4 && (dest[0] == '+' || dest[0] == '-'))
```
Combine this and the case above to remove the redundant expression `dest.size() == 4 && (dest[0] == '+' || dest[0] == '-')`

tautschnig replied:
I was trying to keep the diff smaller, but will now do as suggested.

smowton commented on src/goto-programs/graphml_witness.cpp:
```diff
@@ -47,13 +61,63 @@ void graphml_witnesst::remove_l0_l1(exprt &expr)
+namespace std // NOLINT(readability/namespace)
+{
+template <typename S, typename T>
+struct hash<pair<S, T>> // NOLINT(readability/identifiers)
```
Surely this belongs in util somewhere

tautschnig replied:
Well, it has this single user...

smowton commented on src/goto-programs/graphml_witness.cpp:
```diff
@@ -47,13 +61,63 @@ void graphml_witnesst::remove_l0_l1(exprt &expr)
   const irep_idt &identifier,
   const code_assignt &assign)
 {
+  static std::
```
Quite surprised the lifetime of this cache isn't tied to that of some class, rather than living as long as the process

tautschnig replied:
I'll make it a class member.

smowton commented on src/goto-programs/graphml_witness.cpp:
```diff
@@ -124,18 +188,57 @@ std::string graphml_witnesst::convert_assign_rec(
+      }
+      else
+      {
+        const index_exprt index{assign.lhs(), ops[i]};
```
Invariant that this is an `ID_array_index` or whatever it's called?

tautschnig replied:
No, this is the literal index, not a wrapper around an index. It could be any expression.

smowton commented on src/goto-programs/graphml_witness.cpp:
```diff
@@ -169,22 +272,93 @@ static bool filter_out(
+      graphml[*it].out[*std::next(it)].xml_node = edge;
+    }
+
+    // we do not provide any further details as CPAchecker does not seem to
```
Perhaps link to some appropriate issue or discussion thread?

tautschnig replied:
A link to furious discussions at the SV-COMP meeting in Thessaloniki? :-)  Or maybe https://github.com/sosy-lab/sv-witnesses/issues/12?

smowton commented on src/goto-programs/graphml_witness.cpp:
```diff
@@ -262,73 +438,85 @@ void graphml_witnesst::operator()(const goto_tracet &goto_trace)
+          data_t.data = std::to_string(++thread_id);
+        }
+        else if(
+          !contains_symbol_prefix(
```
Comment on why dynamic objects are excluded?

tautschnig replied:
Comment is in the commit message: dynamic objects aren't part of the input source code, and therefore aren't known to any witness validator.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: ee2c68bf1d83b6606e776dfc14443ef0959186d3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105821202

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 16a59aadf96f1eb65c96afd87c50c37aa4d112e2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/107154317
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.
